pymc_marketing/customer_choice/nested_logit.py

```diff
@@ -0,0 +1,783 @@
+            model = self.make_model(X, F, y)
+            self.model = model
+
+        self.sample_prior_predictive(
+            extend_idata=True, kwargs=sample_prior_predictive_kwargs
+        )
+        self.fit(extend_idata=True, kwargs=fit_kwargs)
+        self.sample_posterior_predictive(
+            extend_idata=True, kwargs=sample_posterior_predictive_kwargs
+        )
+        return self
+
+    def apply_intervention(self, new_choice_df, new_utility_equations=None):
+        """Apply one of two types of intervention.
+
+        The first type of intervention assumes we have a fitted model and
+        just aims to sample from the posterior predictive distribution after
+        adjusting one of more of the models observable attributes and passing
+        in the new_choice_df. The second type of intervention allows that we remove a
+        product entirely from the market place and model the market share which
+        accrues to each product in the adjusted market.
+        """
+        if not hasattr(self, "model"):
+            self.sample()
+        if new_utility_equations is None:
+            new_X, new_F, new_y = self.preprocess_model_data(
+                new_choice_df, self.utility_equations
+            )
+            with self.model:
+                if new_F is None:
+                    pm.set_data({"X": new_X, "y": new_y})
+                else:
+                    pm.set_data({"X": new_X, "W": new_F, "y": new_y})
+                # use the updated values and predict outcomes and probabilities:
+                idata_new_policy = pm.sample_posterior_predictive(
+                    self.idata,
+                    var_names=["p", "likelihood"],
+                    return_inferencedata=True,
+                    extend_inferencedata=False,
+                    random_seed=100,
+                )
+
+            self.intervention_idata = idata_new_policy
+        else:
+            new_X, new_F, new_y = self.preprocess_model_data(
+                new_choice_df, new_utility_equations
+            )
+            new_model = self.make_model(new_X, new_F, new_y)
+            with new_model:
+                idata_new_policy = pm.sample_prior_predictive()
+                idata_new_policy.extend(
+                    pm.sample(
+                        target_accept=0.99,
+                        tune=2000,
+                        idata_kwargs={"log_likelihood": True},
+                        random_seed=101,
+                    )
+                )
+                idata_new_policy.extend(
+                    pm.sample_posterior_predictive(
+                        idata_new_policy, var_names=["p", "likelihood"]
+                    )
+                )
+
+            self.intervention_idata = idata_new_policy
+
+        return idata_new_policy
+
+    @staticmethod
+    def calculate_share_change(idata, new_idata):
```

Comment:
type hints missing and parameters in docstings

Reply:
Done